src/persistent_state.py

```diff
@@ -0,0 +1,168 @@
+from typing import Iterable
+
+from config import JSON_DATA_TYPE, bot_state_file
+
+# Global, persistent state of the bot. Not to be accessed directly. Instead, use the
+# getter and setter methods below.
+_bot_state = {}
+
+
+def load_state_from_disk() -> None:
+    """Read the bot state from disk and store it in memory."""
+    global _bot_state
+
+    with open(bot_state_file) as f:
+        bot_state_str = f.read()
+
+        if bot_state_str:
+            # Load the JSON representation of the bot's state and convert it to a Python dict so
+            # it can be easily manipulated.
+            _bot_state = json.loads(bot_state_str)
+
+
+def set_state(path: Iterable[str], value: JSON_DATA_TYPE) -> None:
+    """Modifies the bot's state in memory, then dumps the modified state to disk.
+
+    This function should not be made async! We do not want to yield from it before we update
+    the bot's state on disk. Otherwise, this data model is thread-safe as long as there are
+    never >1 simultaneous writes to disk, which python will not let happen without
+    multi-processing.
+
+    Args:
+        path: The path to store the state under. For example, if storing the settings for a
+            particular channel in a guild, you might set `path` to
+            ["channel_settings", <guild_id>, <channel_id>, <setting_id>].
+        value: The value to store. Only data types that can be serialised to JSON may be
+            used, as the state is backed by a JSON store.
+    """
+    global _bot_state
+
+    # We don't want to modify the given path directly, so we make a copy here.
+    path = list(path)
+
+    # Check if we need to traverse the bot state via the given path.
+    if path:
+        # Start at the top of the dict
+        current_path = _bot_state
+
+        # Iterate down the provided path until we reach the second from the last entry in the path.
+        # We create a new dictionary and enter it if we hit a dead end. We'll use the final entry in
+        # `path` as the field to store `value` under.
+        #
+        # Therefore, a `path` list of ["settings", "guild_id_123", "submission_channel"] and a `value`
+        # of "channel_id_456" would yield a bot_state dictionary of:
+        #
+        # {"settings": {"guild_123": {"submission_channel": "channel_id_456"}}}
+        key = path.pop()
+        for pathname in path:
+            current_path = current_path.setdefault(pathname, {})
+
+        # `current_path` is now set to the dictionary where we're like to place our `key` and our `value`.
+        # As `current_path` is just a reference to a dictionary inside `bot_state`, this modifies
+        # `bot_state` as well.
+        current_path[key] = value
+    else:
+        # If no path was provided, just override all bot state with the given value.
+        # The given value must be a dict.
+        if not isinstance(value, dict):
+            raise Exception("Attempted to override all bot state with a non-dict type")
```

Comment:
```suggestion
            raise Exception("Attempted to override entire bot state with a non-dict type")
```
Just because I like the sound of it better 8^)